File: client/javascript/src/mpc/filter.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {Invocation, Invoker} from "@/mpc/invoker";
import {Context} from "@/prsim";
import {Ark, ServiceLoader, SPI, spi} from "@/macro";

export const PROVIDER: string = "PROVIDER";
export const CONSUMER: string = "CONSUMER";

@spi("*")
export abstract class Filter {

    /**
     * Invoke the next filter.
     * @param ctx
     * @param invoker
     * @param invocation
     */
    abstract invoke(ctx: Context, invoker: Invoker<any>, invocation: Invocation): Promise<any>
}

/**
 * Composite the filter spi providers as a invoker.
 * @param invoker
 * @param pattern
 */
export function composite(invoker: Invoker<any>, pattern: string): Invoker<any> {
    const filters = ServiceLoader.load(Filter).list().filter(filter => {
        const spi = Ark.metadata(SPI, filter?.constructor, filter?.constructor);
        return !spi || spi.pattern == "" || spi?.pattern == pattern;
    }).sort((p, n): number => {
        const pp = Ark.metadata(SPI, p?.constructor, p?.constructor)?.priority || 0;
        const np = Ark.metadata(SPI, n?.constructor, n?.constructor)?.priority || 0;
        return pp == np ? 0 : (pp < np ? -1 : 1);
    });
    return composite0(invoker, filters);
}

function composite0(invoker: Invoker<any>, filters: Filter[]): Invoker<any> {
    let last = invoker;
    for (let index = filters.length - 1; index >= 0; index--) {
        let filter = filters[index];
        let next = last;
        last = new class implements Invoker<any> {
            run(ctx: Context, invocation: Invocation): Promise<any> {
                return filter.invoke(ctx, next, invocation);
            }
        }
    }
    return last;
}

File: client/golang/mpc/filter.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package mpc

import (
	"context"
	"github.com/opendatav/mesh/client/golang/macro"
	"sort"
)

const (
	// PROVIDER Flag.
	PROVIDER = "PROVIDER"
	// CONSUMER Flag.
	CONSUMER = "CONSUMER"
)

var IFilter = (*Filter)(nil)

type Filter interface {

	// Invoke the next filter.
	// invoker    Service invoker.
	// invocation Service invocation
	Invoke(ctx context.Context, invoker Invoker, invocation Invocation) (interface{}, error)

	macro.SPI
}

type filters []Filter

func (that filters) Len() int {
	return len(that)
}

func (that filters) Less(x, y int) bool {
	return that[x].Att().Priority < that[y].Att().Priority
}

func (that filters) Swap(x, y int) {
	tmp := that[y]
	that[y] = that[x]
	that[x] = tmp
}

// Composite the filter spi providers as an invoker.
// invoker Delegate invoker.
// pattern Filter pattern.
func composite(invoker Invoker, pattern string) Invoker {
	var filters filters
	x := macro.Load(IFilter).List()
	for _, filter := range x {
		filter, ok := filter.(Filter)
		if !ok || filter.Att().Pattern != pattern {
			continue
		}
		filters = append(filters, filter)
	}
	sort.Sort(filters)
	return composite0(invoker, filters)
}

// Composite the filter spi providers as a invoker.
// invoker Delegate invoker.
// filters Filter list.
func composite0(invoker Invoker, filters []Filter) Invoker {
	last := invoker
	for i := len(filters) - 1; i >= 0; i-- {
		filter := filters[i]
		next := last
		var x InvokerFn = func(ctx context.Context, invocation Invocation) (interface{}, error) {
			return filter.Invoke(ctx, next, invocation)
		}
		last = x
	}
	return last
}
